src/server/page_server.cpp

```diff
@@ -913,22 +980,95 @@ page_server::disconnect_all_tran_servers ()
   er_log_debug (ARG_FILE_LINE, "disconnect_all_tran_server: All connections have been disconnected.\n");
 }
 
+void
+page_server::disconnect_all_follower_page_servers ()
+{
+  /*
+   * TODO it should request to disconnect to all followers and wait until it's done.
+   * It will be addressed when follower has a way to re-start the catch-up with another followee.
+   * For now, this waits until just all the catch-up jobs in progress are done.
+   */
+  constexpr auto millis_20 = std::chrono::milliseconds { 20 };
+
+  er_log_debug (ARG_FILE_LINE, "Wait until all follower connections are disconnected.\n");
+
+  {
+    auto ulock_conn_vec = std::unique_lock { m_follower_conn_vec_mutex };
+
+    while (!m_follower_conn_vec_cv.wait_for (ulock_conn_vec, millis_20, [this]
+    {
+      return m_follower_conn_vec.empty ();
+      }));
+
+    assert (m_follower_conn_vec.empty());
+  }
+
+  {
+    auto lockg_disc = std::lock_guard { m_follower_disc_mutex };
+    if (m_follower_disc_future.valid ())
+      {
+	m_follower_disc_future.get (); // wait until it's done if there is an on-going disconnection.
+      }
+  }
+  er_log_debug (ARG_FILE_LINE, "All follower connections have been disconnected.\n");
+}
+
 void
 page_server::disconnect_followee_page_server (const bool with_disc_msg)
 {
-  auto lockg = std::lock_guard <std::mutex> (m_followee_conn_mutex);
+  auto lockg = std::lock_guard { m_followee_conn_mutex };
   if (m_followee_conn != nullptr)
     {
+      auto channel_id = m_followee_conn->get_channel_id ();
       if (with_disc_msg)
 	{
-	  //TODO send the disconnect msg
+	  m_followee_conn->push_request (follower_to_followee_request::SEND_DISCONNECT_MSG, std::string ());
 	}
       m_followee_conn.reset (nullptr);
 
-      er_log_debug (ARG_FILE_LINE, "The followee page server has been disconnected.\n");
+      er_log_debug (ARG_FILE_LINE, "The followee page server has been disconnected. channel id: %s\n", channel_id.c_str ());
     }
 }
 
+void
+page_server::disconnect_follower_page_server_async (const follower_connection_handler *conn)
+{
+  auto ulock_conn_vec = std::unique_lock { m_follower_conn_vec_mutex };
+
+  auto it = std::find_if (m_follower_conn_vec.begin (), m_follower_conn_vec.end (),
+			  [&conn] (auto & conn_uptr)
+  {
+    return conn_uptr.get () == conn;
+  });
+
+  if (it == m_follower_conn_vec.cend ())
+    {
+      // It's already been disconnected or in progress by another thread.
+      return;
+    }
+
+  auto disconnecting_conn_uptr = std::move (*it);
+  m_follower_conn_vec.erase (it);
+  m_follower_conn_vec_cv.notify_one ();
```

Comment:
The way I see it, it's resolved.
> after notify one the control moves and the function page_server::disconnect_all_follower_page_servers is executed completely

It's not the case, because `disconnect_all_follower_page_servers` can't finish. When checking if the vector is empty, it acquires the lock of `m_follower_conn_vec_mutex`, but it's blocked if it's not released in this function. And, if the lock is released, it acquires the lock of `m_follower_disc_mutex`, which is   held by `disconnect_follower_page_server_async`, so it's blocked. Note that the lock of `m_follower_conn_vec_mutex` is unlocked  means that `m_follower_disc_mutex` is held in `disconnect_follower_page_server_async`.